c10/core/DispatchKeySet.h

```diff
@@ -8,6 +7,28 @@
 
 namespace c10 {
 
+struct FunctionalityOffsetAndMask {
+    // empty constructor shouldn't be used; only needed to initialize
+    // the array before populating it.
+    FunctionalityOffsetAndMask() {}
+    FunctionalityOffsetAndMask(uint16_t offset, uint16_t mask) :
+        offset(offset), mask(mask) {}
+    // This needs to big enough to cover the size of the operator table.
+    uint16_t offset;
+    // This mask needs to be big enough to mask all of the backend bits.
+    // We probably don't ever want to have more than 16 backend bits, so uint16_t should be enough.
+    uint16_t mask;
+};
+
+C10_API std::array<FunctionalityOffsetAndMask, num_functionality_keys>
+initializeFunctionalityOffsetsAndMasks();
+
+C10_ALWAYS_INLINE static std::array<FunctionalityOffsetAndMask, num_functionality_keys>&
```

Comment:
are you sure you want this to return a non-const reference? seems like a thread safety footgun.

Reply:
this should definitely be a const reference, thanks!